node/lib/DeviceAccessor.ts

```diff
@@ -0,0 +1,241 @@
+
+        if (typeof removeListenerMethod !== "function") {
+            throw new TypeError("Property '" + propertyName + "' notifier removal " +
+                "for interface " + interfaceName + " not implemented by device.");
+        } else {
+            removeListenerMethod.call(deviceInterface, propertyName, callback);
+        }
+    }
+
+    /**
+     * Invokes a method on a device, using a specified interface.
+     *
+     * @param {IDevice} device  Device instance obtained from a translator
+     * @param {(string | DeviceInterface)} interfaceName  Name of the interface used
+     *     to access the device, or a DeviceInterface object
+     * @param {string} methodName  Name of the method to invoke
+     * @param {any[]} args  Arguments to pass to the method; the number and types of
+     *     arguments are expected to conform to the interface definition of the method
+     *     and the JSON schemas for each method argument in the interface.
+     * @returns {Promise<any>} Value returned by the method, or undefined if the
+     *     method has a void return type; the value is expected to conform to the
+     *     JSON schema for the method return value as specified by the interface
+     */
+    public static async invokeMethodAsync(
+            device: IDevice,
+            interfaceName: string | DeviceInterface,
+            methodName: string,
+            args: any[]): Promise<any> {
+        let deviceInterface = DeviceAccessor.getDeviceInterface(device, interfaceName);
+        DeviceAccessor.validateMemberName(methodName);
+        if (!Array.isArray(args)) {
+            throw new TypeError("Args argument must be an array.");
+        }
+
+        let method: any = deviceInterface[methodName];
+        if (typeof method !== "function") {
+            throw new TypeError("Method '" + methodName + "' " +
+                "for interface " + interfaceName + " not implemented by device.");
+        } else {
+            let result = method.apply(deviceInterface, args);
+            if (typeof result === "object" && typeof result.then === "function") {
+                return await result;
+            } else {
+                return result;
+            }
+        }
+    }
+
+    /**
+     * Check for an `as` method on the device, and if found use it to request an object
+     * that implements the specified interface.
+     */
+    private static getDeviceInterface(
+            device: IDevice, interfaceName: string | DeviceInterface): {[key: string]: any} {
+        if (typeof device !== "object") {
+            throw new TypeError("Device argument must be an object.");
+        }
+
+        if (typeof device.as !== "function") {
+            // The device doesn't implement an "as" method, so it is assumed to implement
+            // all interface methods directly.
+            return <{[key: string]: any}> device;
+        }
+
+        if (typeof interfaceName !== "string") {
```

Comment:
In JavaScript, it's harder to check whether an object is an instance of a class: `typeof` will just return `"object"`. Since it should only ever be one or the other, this `!== "string"` check is sufficient.

---
In reply to: [71899338](https://github.com/opent2t/opent2t/pull/1#discussion_r71899338) [](ancestors = 71899338)